crates/gooey-widgets/src/input.rs

```diff
@@ -0,0 +1,243 @@
+            renderer.fill_rect(renderer.size().into_signed().into());
+            self.input.value.map_ref(|label| {
+                // TODO use the width
+                let metrics: TextMetrics<Px> = renderer.measure_text(label, None);
+
+                renderer.fill.color = control_text_color(self.state);
+                let render_size = renderer.size().into_signed();
+                renderer.draw_text(
+                    label,
+                    (Point::new(
+                        render_size.width - metrics.size.width,
+                        render_size.height + metrics.ascent,
+                    )) / 2,
+                    None,
+                );
+            });
+        }
+
+        fn mouse_down(&mut self, _event: MouseEvent, context: &mut dyn AnyRasterContext) {
+            self.state = State::Active;
+            // TODO position
+            context.invalidate();
+        }
+
+        fn cursor_moved(&mut self, event: MouseEvent, context: &mut dyn AnyRasterContext) {
+            let changed = event.position.is_some() != (self.state == State::Hover);
+            if changed {
+                if event.position.is_some() {
+                    self.state = State::Hover;
+                } else {
+                    self.state = State::Normal;
+                }
+                context.invalidate();
+            }
+        }
+
+        fn key_down(&mut self, event: KeyEvent, context: &mut dyn AnyRasterContext) {
+            'outer: {
+                if let Some(key) = event.virtual_keycode {
+                    match key {
+                        _ if (VirtualKeyCode::A..=VirtualKeyCode::Z).contains(&key) => {
+                            self.value.push(
+                                char::from_u32(
+                                    (key as u32) - (VirtualKeyCode::A as u32) + ('a' as u32),
+                                )
+                                .expect("valid key"),
+                            );
+                        }
+                        VirtualKeyCode::Space => self.value.push(' '),
+                        VirtualKeyCode::Backspace => self.value.truncate(self.value.len() - 1),
+                        VirtualKeyCode::Underline => self.value.push('_'),
+                        VirtualKeyCode::Minus => self.value.push('-'),
+                        _ => {
+                            eprintln!("unknown: `{key:?}`");
+                            break 'outer;
+                        }
+                    }
+                    if let Value::Dynamic(value) = self.input.value {
+                        value.set(self.value.clone());
+                    }
+                    if let Some(on_update) = self.input.on_update.as_mut() {
+                        on_update.invoke(self.value.clone());
+                    }
+                } else {
+                    eprintln!("Missing virtual keycode: {}", event.scan_code);
```

Comment:
I noticed that virtual keycodes are very much not useful when one wants to do input, e.g., `ä` just doesn't trigger any virtual keycode, not even an incorrect one.

Reply:
I'll check this again with the new winit

Reply:
We need the `receive_character` event from appit to receive text input. In general, virtual keycodes should only be used when they represent keys that don't generate characters when pressed.

Reply:
Ok, now that you've pointed out winit changed everything, this comment is outdated. But this is still the real solution: don't use virtual key codes unless. you *need* to, which is almost never.